src/openrc/rc.c

```diff
@@ -139,10 +139,13 @@ cleanup(void)
 		rc_logger_close();
 	}
 
-	while (p1) {
-		p2 = LIST_NEXT(p1, entries);
-		free(p1);
-		p1 = p2;
+	LIST_FOREACH_SAFE(p, &service_pids, entries, tmp) {
+		LIST_REMOVE(p, entries);
```

Comment:
Are signals blocked by the time we get here? (sorry not able to check quickly right now)
This is a good simplification however, list_foreach_safe is perfect for this.